ice-adapter/src/main/java/com/faforever/iceadapter/ice/GameSession.java

```diff
@@ -86,36 +87,53 @@ public static void setIceServers(List<Map<String, Object>> iceServersData) {
             return;
         }
 
-        //TODO: support multiple ice servers
+        // For caching RTT to a given host (the same host can appear in multiple urls)
+        LoadingCache<String, CompletableFuture<OptionalDouble>> hostRTTCache = CacheBuilder.newBuilder().build(
+                new CacheLoader<String, CompletableFuture<OptionalDouble>>() {
+                    @Override
+                    public CompletableFuture<OptionalDouble> load(String host) throws Exception {
+                        return PingWrapper.getLatency(host, IceAdapter.PING_COUNT)
+                                .thenApply(OptionalDouble::of)
+                                .exceptionally(ex -> OptionalDouble.empty());
+                    }
+                }
+        );
 
         for(Map<String, Object> iceServerData : iceServersData) {
             IceServer iceServer = new IceServer();
 
-            if(iceServerData.containsKey("username")) {
+            if (iceServerData.containsKey("username")) {
                 iceServer.setTurnUsername((String) iceServerData.get("username"));
             }
-            if(iceServerData.containsKey("credential")) {
+            if (iceServerData.containsKey("credential")) {
                 iceServer.setTurnCredential((String) iceServerData.get("credential"));
             }
 
-            if(iceServerData.containsKey("urls") && iceServerData.get("urls") instanceof List) {
-                ((List<String>) iceServerData.get("urls")).stream()
-                        .map(iceServerUrlPattern::matcher)
+            if (iceServerData.containsKey("urls")) {
+                List<String> urls;
+                Object urlsData = iceServerData.get("urls");
+                if (urlsData instanceof List) {
+                    urls = (List<String>) urlsData;
+                } else {
+                    urls = Collections.singletonList((String) iceServerData.get("url"));
+                }
+
+                urls.stream()
+                        .map(IceServer.urlPattern::matcher)
                         .filter(Matcher::matches)
                         .forEach(matcher -> {
-                            TransportAddress address = new TransportAddress(matcher.group("host"), matcher.group("port") != null ? Integer.parseInt(matcher.group("port")) : 3478, matcher.group("protocol").equals("stun") ? Transport.UDP : Transport.parse(matcher.group("transport")));
+                            String host = matcher.group("host");
+                            int port = matcher.group("port") != null ? Integer.parseInt(matcher.group("port")) : 3478;
```

Comment:
Not sure if that's better, but to avoid repeating the `matcher.group("port")` you could do:
```int port = OptionalInt.ofNullable(matcher.group("port")).orElse(3478);```

Reply:
Yea I like that way better